File: rust/advent-of-code/src/four.rs
use std::{
    fs::File,
    io::{BufRead, BufReader, Lines},
};

fn read_pairs_from_file() -> Lines<BufReader<File>> {
    let file = match File::open("src/inputs/four.txt") {
        Ok(file) => file,
        Err(error) => panic!("Cannot open day 4 input file, error: {:?}", error),
    };

    let reader = BufReader::new(file);
    return reader.lines();
}

pub fn count_overlap_pair() -> i32 {
    let pairs = read_pairs_from_file();
    let mut counter = 0;

    for pair in pairs {
        match pair {
            Ok(pair) => {
                let mut extreme_points: Vec<i32> = vec![];
                let ranges: Vec<&str> = pair.split(",").collect();
                for range in ranges {
                    let points: Vec<&str> = range.split("-").collect();
                    for point in points {
                        let value: i32 = point.parse().unwrap();
                        extreme_points.push(value);
                    }
                }

                if is_overlap_ranges(
                    *extreme_points.get(0).unwrap(),
                    *extreme_points.get(1).unwrap(),
                    *extreme_points.get(2).unwrap(),
                    *extreme_points.get(3).unwrap(),
                ) {
                    counter += 1;
                }
            }
            Err(error) => panic!("Cannot read with error {:?}", error),
        };
    }

    return counter;
}

fn is_overlap_ranges(start1: i32, end1: i32, start2: i32, end2: i32) -> bool {
    return (start1 <= start2 && end1 >= end2) || (start1 >= start2 && end1 <= end2);
}

File: golang/advent-of-code/day4/four.go
package day4

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func ReadAssignmentFromFile() []string {
	file, err := os.Open("day4/assignment.txt")
	if err != nil {
		fmt.Println("Cannot find/open assignment.txt")
	}

	reader := bufio.NewReader(file)
	var result = []string{}

	for {
		line, _, err := reader.ReadLine()
		if err != nil {
			break
		}

		result = append(result, string(line))
	}
	return result
}

func CounOverlapPair(inputs []string) int32 {
	var counter int32 = 0
	for _, input := range inputs {
		var extremePoints = []int32{}

		for _, pointRange := range strings.Split(input, ",") {
			for _, point := range strings.Split(pointRange, "-") {
				value, err := strconv.Atoi(point)
				if err != nil {
					panic(err)
				}
				extremePoints = append(extremePoints, int32(value))
			}
		}

		if isFullOverlap(
			extremePoints[0],
			extremePoints[1],
			extremePoints[2],
			extremePoints[3],
		) {
			counter++
		}
	}

	return counter
}

func CountPartialOverlapPair(inputs []string) int32 {
	var counter int32 = 0
	for _, input := range inputs {
		var extremePoints = []int32{}

		for _, pointRange := range strings.Split(input, ",") {
			for _, point := range strings.Split(pointRange, "-") {
				value, err := strconv.Atoi(point)
				if err != nil {
					panic(err)
				}
				extremePoints = append(extremePoints, int32(value))
			}
		}

		if isPartialOverlap(
			extremePoints[0],
			extremePoints[1],
			extremePoints[2],
			extremePoints[3],
		) {
			counter++
		}
	}

	return counter
}

func isFullOverlap(start1, end1, start2, end2 int32) bool {
	return (start1 <= start2 && end1 >= end2) ||
		(start2 <= start1 && end2 >= end1)
}

func isPartialOverlap(start1, end1, start2, end2 int32) bool {
	return (start1 <= start2 && end1 >= start2) ||
		(start1 <= end2 && end1 >= end2) ||
		(start2 <= start1 && end2 >= start1) ||
		(start2 <= end1 && end2 >= end1)
}
